File: app/javascript/discovery/wa_sos_funnel_filters.js
const ADDRESS_COLUMN = "Office Address"

// Keep in sync with Discovery::Sources::WaSos::Cities
const SOUTHERN_WA = "Southern WA"
const INDIVIDUAL_CITIES = [
  "Vancouver",
  "Camas",
  "Washougal",
  "Ridgefield",
  "Battle Ground",
  "Hockinson",
  "Brush Prairie",
  "Woodland"
]

export function extractCity(address) {
  if (!address || !/,\s*WA\b/i.test(address)) return null

  const beforeState = address.split(/,\s*WA\b/i)[0]
  const segments = beforeState.split(",").map((segment) => segment.trim()).filter(Boolean)
  return segments[segments.length - 1] || null
}

function matchCities(city) {
  const normalized = (city || "").trim()
  if (!normalized) return []
  return normalized.localeCompare(SOUTHERN_WA, undefined, { sensitivity: "accent" }) === 0
    ? INDIVIDUAL_CITIES
    : [normalized]
}

function filterByCity(rows, city) {
  if (!city) return rows

  const targets = matchCities(city)
  return rows.filter((row) => {
    const extracted = extractCity(row[ADDRESS_COLUMN])
    return (
      extracted &&
      targets.some(
        (target) => extracted.localeCompare(target, undefined, { sensitivity: "accent" }) === 0
      )
    )
  })
}

function filterByBusinessName(rows, query) {
  const normalized = (query || "").trim().toLowerCase()
  if (!normalized) return rows

  return rows.filter((row) => {
    const name = String(row["Business Name"] || "").toLowerCase()
    return name.includes(normalized)
  })
}

// Post-fetch funnel — add more filters here before persisting to DB (Phase 2).
export function applyFunnelFilters(rows, { city, businessName } = {}) {
  let filtered = rows
  filtered = filterByCity(filtered, city)
  filtered = filterByBusinessName(filtered, businessName)
  return filtered
}

File: app/services/discovery/sources/wa_sos_funnel_filters.rb
# frozen_string_literal: true

module Discovery
  module Sources
    # Post-fetch funnel — narrows SOS rows before display and (Phase 2) DB import.
    # Add more filters here as the funnel grows; city is the first.
    class WaSosFunnelFilters
      ADDRESS_COLUMN = Discovery::Sources::WaSosCsvParser::OFFICE_ADDRESS_COLUMN

      def self.apply(rows, city: nil)
        filtered = rows
        filtered = filter_by_city(filtered, city) if city.present?
        filtered
      end

      def self.extract_city(address)
        return nil if address.blank?
        return nil unless address.match?(/,\s*WA\b/i)

        before_state = address.split(/,\s*WA\b/i).first.to_s
        segments = before_state.split(",").map(&:strip).reject(&:blank?)
        segments.last
      end

      def self.filter_by_city(rows, city)
        targets = Discovery::WaSosCities.match_cities(city)
        rows.select { |row| matches_cities?(row, targets) }
      end

      def self.matches_cities?(row, cities)
        extracted = extract_city(row[ADDRESS_COLUMN])
        return false if extracted.blank?

        cities.any? { |target| extracted.casecmp?(target.to_s.strip) }
      end

      def self.matches_city?(row, city)
        matches_cities?(row, [city])
      end
    end
  end
end
